csrc/scheduler/registry_utils.cpp

```diff
@@ -442,60 +442,88 @@ bool reductionInterferingView(
 
-  // Make sure groups are disjoint based on view
-
-  auto disjoint_rfactor_sets = scheduler_utils::disjointLogicalSets(fusion);
-  auto disjoint_set_information = scheduler_utils::getDisjointLogicalSetsOf(
-      fusion, reduction_reference, disjoint_rfactor_sets);
-
-  // Convert id's in groups to disjoint_set_ids of disjoint_set_information
-  std::vector<std::vector<int64_t>> disjoint_groups;
-
-  for (const auto& group : groups) {
-    std::vector<int64_t> disjoint_id_sets;
-    for (auto id : group) {
-      auto find_it = std::find(
-          reduction_reference->getLogicalDomain().begin(),
-          reduction_reference->getLogicalDomain().end(),
-          id);
-      NVF_ERROR(
-          find_it != reduction_reference->getLogicalDomain().end(),
-          "Issue with view analysis on reduction like schedule, with reference: ",
-          reduction_reference->toString());
-      auto logical_pos = std::distance(
-          reduction_reference->getLogicalDomain().begin(), find_it);
-      NVF_ERROR(
-          logical_pos <
-              (int64_t)disjoint_set_information.disjoint_set_ids.size(),
-          "Error computing disjoint group on the logical domain of ",
-          reduction_reference->toString());
-      disjoint_id_sets.push_back(
-          disjoint_set_information.disjoint_set_ids[logical_pos]);
+  // Check if any view op could merge IDs in different coalesced_ids.
+  // (1) Build the exact graph
+  IdModel id_model(fusion, false, false, false);
+  id_model.buildExactGraph();
+  const ValGraph& graph = id_model.idGraph(IdMappingMode::EXACT);
+  auto disjoint_sets = graph.disjointValSets();
+  for (auto view : ir_utils::getViewOps(fusion)) {
+    auto tv = view->out();
+    // keep track of split IDs, each unordered_set in the vector stores
+    // connected split IDs. For example, if we have:
+    // I0 -> I1, I2
+    // I3 -> I4, I5
+    // I5 -> I6, I7
+    // There are 2 groups: [I0, I1, I2] and [I3, I4, I5, I6, I7]
+    // when a merge op is found, e.g. I9 = I8 * I7, we need to map
+    // I9 with I8 and all the IDs in [I3, I4, I5, I6, I7].
+    std::vector<std::unordered_set<IterDomain*>> ids_in_connected_splits;
+    for (auto expr : StmtSort::getExprsTo(
+             {tv->getLogicalDomain().begin(), tv->getLogicalDomain().end()})) {
+      if (auto merge = dynamic_cast<Merge*>(expr)) {
+        auto id1 = merge->inner();
+        auto id2 = merge->outer();
+        disjoint_sets.mapEntries(id1, merge->out());
+        disjoint_sets.mapEntries(id2, merge->out());
+        // if merged id comes from a set of connected splits, map it with all
+        // other IDs in the set
+        for (auto merged_id : {id1, id2}) {
+          auto it = std::find_if(
+              ids_in_connected_splits.begin(),
+              ids_in_connected_splits.end(),
+              [&merged_id](const auto& id_set) {
+                return id_set.count(merged_id);
```

Comment:
I don't think you can just search by IterDomain. Remember we are traversing a graph where each node is a group of iter domains.

Reply:
A little confused here. we are traversing from a tv's root domain to its logical domain using `StmtSort::getExprsTo(
             {tv->getLogicalDomain().begin(), tv->getLogicalDomain().end()})`. So each node is an iter domain.
For example, a fusion with
```
  auto tv2 = reshape(tv1, {12, 20 ,3}, {3, 4, 5, 2, 6});
  auto tv3 = sum(tv1, {-1});
```
From tv2's root to its logical domain, we have:
![image](https://github.com/NVIDIA/Fuser/assets/116412316/3da1443c-9f10-4329-aeb7-aee465455cf1)
In this graph, each node is an iter domain.
```
 root domain : (iS9{i0}rf, iS12{i2}rf, iS17{i3}rf)
  Outer split: iS9{i0}rf by factor 3 -> iS10{3}rf, iS11{( ceilDiv(i0, 3) )}rf
  Outer split: iS12{i2}rf by factor 5 -> iS13{5}rf, iS14{( ceilDiv(i2, 5) )}rf
  Outer split: iS14{( ceilDiv(i2, 5) )}rf by factor 2 -> iS15{2}rf, iS16{( ceilDiv(( ceilDiv(i2, 5) ), 2) )}rf
  Merge: iS16{( ceilDiv(( ceilDiv(i2, 5) ), 2) )}rf and iS17{i3}rf -> iS18{( ( ceilDiv(( ceilDiv(i2, 5) ), 2) ) * i3 )}rf
 logical domain : (iS10{3}rf, iS11{( ceilDiv(i0, 3) )}rf, iS13{5}rf, iS15{2}rf, iS18{( ( ceilDiv(( ceilDiv(i2, 5) ), 2) ) * i3 )}rf)
```
After visiting the first 3 splits, `ids_in_connected_splits` has 2 sets:
```
set1 from i0: iS10{3}rf iS11{( ceilDiv(i0, 3) )}rf iS9{i0}rf 
set2 from i2: iS15{2}rf iS16{( ceilDiv(( ceilDiv(i2, 5) ), 2) )}rf iS13{5}rf iS14{( ceilDiv(i2, 5) )}rf iS12{i2}rf 
```
When visit the merge expr, this line of code will find `iS16{( ceilDiv(( ceilDiv(i2, 5) ), 2) )}rf` from `set2` in `ids_in_connected_splits`. Then, all other IDs in `set2` are mapped with output of the merge expr.

Reply:
oh, great point. I addd 3 additional tests testing multiple reshape ops. Revised to use ValGroup.
```
  // Here we can't directly find IterDomain since they may be in different
  // reshape ops. ValGroup groups IDs in different tvs.
  auto findIdSet = [&](IterDomain* id) {
    auto id_vgroup = graph.toGroup(id);
    return std::find_if(
        ids_in_connected_splits.begin(),
        ids_in_connected_splits.end(),
        [&](const std::unordered_set<IterDomain*>& id_set) {
          return std::any_of(
              id_set.begin(), id_set.end(), [&](IterDomain* set_id) {
                return graph.toGroup(set_id) == id_vgroup;
              });
        });
  };
```

Reply:
For example when visit the merge of `i2/5/2` in `reshape-2`, the code is able to find `i2/5` and `i2` in `reshape-1`
![image](https://github.com/NVIDIA/Fuser/assets/116412316/5c25e874-14d4-4f36-9489-c1111d86a6f2)

Reply:
Originally you had code to filter out reshapes before reductions. Don't we still want to do that?

Reply:
we can add that and also skip the following checks if there is no merge detetected in any of the reshapes. Other than that, does the current method looks good?

Reply:
Yes, let's add it back.

Reply:
Thanks! Added.

Reply:
This won't work if there are reshapes both before and after reductions, right?

Reply:
I think it still works fine. For example, this fusion has reshape before and after reduction,
![image](https://github.com/user-attachments/assets/7b04704f-fb59-4092-a5ab-95e72eccdbe2)
In this fusion, the first reshape is not processed since it is used by the reduction. When processing the merge op in the 2nd reshape, `i2/5/2` is mapped to `i2/5` and all other `i2/5` domains from `t3, t4, and t5`. The code can still detect the merge of an iteration domain and a reduction domain. Do  you have an example where the code may fail? The tests include several other cases but all work fine.

Reply:
How about this?

```
auto tv0 = [I0, I1]; // input
auto tv1 = reshape [I0, I1] -> [I0/X, X*I1/Y, Y]
auto tv2 = sum(tv1, {1}) // output

auto tv3 = reshape [I0, I1] -> [I0/X, X*I1/Y, Y] // output
```

The reshape for `tv3` is unused by the reduction. The split-followed-by-merge pattern makes all domains grouped together, thus this pattern will be considered interfering. But should it be?

Reply:
You are right. This fusion shouldn't be rejected by `reductionInterferingView`. The output of the merge op and one of the reduction tv's root ID are in the same valGroup. It woun't cause the merge of an iter dim with a redu dim when replayed on the reduction tv.
To avoid this, the algorithm should be modified to avoid mapping the merged Ids if its output and one of the reduction tv's root ID are in the same valGroup.

**Code change**
```
  // val group for reduction tv's root IDs
  std::unordered_set<ValGroup> ref_root_groups;
  for (auto id : reduction_reference->getMaybeRootDomain()) {
    ref_root_groups.insert(graph.toGroup(id));
  }

      if (auto merge = dynamic_cast<Merge*>(expr)) {
        // if the output of the merge op and one of the reduction tv's root ID
        // are in the same valGroup. It won't cause the merge of an iter dim
        // with a redu dim when replayed on the reduction tv. We can skip this
        // transform, further transforms using this ID can still be captured and
        // mapped since it it a root ID of the reduction tv. See tests
        // GpuViewTest.SplitMergeReductionSplitMerge and
        // GpuViewTest.SplitMergeReductionSplitMergeSplitMerge
        if (ref_root_groups.count(graph.toGroup(merge->out())) > 0) {
          continue;
        }
```
**New Tests**
(1) GpuViewTest.SplitMergeReductionSplitMerge
```
  // root domain : (i0, i2)
  // logi domain : (3, i0/3, 4, i2/4)
  auto tv2 = reshape(tv1, {12, 20}, {3, 4, 4, 5});
  // root domain : (3, i0/3, 4, i2/4)
  // logi domain : (3, i0/3*4, i2/4)
  auto tv3 = reshape(tv2, {3, 4, 4, 5}, {3, 16, 5});
  // root domain : (3, i0/3*4, i2/4)
  auto tv4 = sum(tv3, {1});
  // root domain : (i0, i2)
  // logi domain : (3, i0/3, 4, i2/4)
  auto tv5 = reshape(tv1, {12, 20}, {3, 4, 4, 5});
  // root domain : (3, i0/3, 4, i2/4)
  // logi domain : (3, i0/3*4, i2/4)
  // merge transform generates `i0/3*4` which is in the same ValGroup as tv4's
  // root ID `i0/3*4`, it is not interfering reduction.
  auto tv6 = reshape(tv5, {3, 4, 4, 5}, {3, 16, 5});
```
Not rejected by `reductionInterferingView`, however, it triggers a bug, see #2593. Marked as skipped.
(2) GpuViewTest.SplitMergeReductionSplitMergeSplitMerge
Modified from `(1)` by adding another split-merge, rejected by `reductionInterferingView` due to the last merge.
![image](https://github.com/user-attachments/assets/cb21cf82-268a-4676-806d-63e3b50b0de1)

Reply:
Did you actually change the code? I don't see anything like `ref_root_groups`.

Reply:
oops, local change it not pushed to remote. Updated.
since #2593 is fixed, the tes case is not skipped.